salt/utils/state.py

```diff
@@ -0,0 +1,155 @@
+                        if not onfail_handled:
+                            continue
+                        for fdata in fchunk:
+                            if not isinstance(fdata, dict):
+                                continue
+                            for knob, fvalue in six.iteritems(fdata):
+                                if knob != 'onfail':
+                                    continue
+                                for freqs in fvalue:
+                                    for fmod, fid in six.iteritems(freqs):
+                                        if not (
+                                            fid == st['__id__'] and
+                                            fmod == st.get('state', fmod)
+                                        ):
+                                            continue
+                                        onfails.append((fstate, mod_, fchunk))
+    return onfails
+
+
+def check_onfail_requisites(state_id, state_result, running, highstate):
+    '''
+    When a state fail and is part of a highstate, check
+    if there is onfail requisites.
+    When we find onfail requisites, we will consider the state failed
+    only if at least one of those onfail requisites also failed
+
+    Returns:
+
+        True: if onfail handlers suceeded
+        False: if one on those handler failed
+        None: if the state does not have onfail requisites
+
+    '''
+    nret = None
+    if (
+        state_id and state_result and
+        highstate and isinstance(highstate, dict)
+    ):
+        onfails = search_onfail_requisites(state_id, highstate)
+        if onfails:
+            for handler in onfails:
+                fstate, mod_, fchunk = handler
+                for rstateid, rstate in six.iteritems(running):
+                    if '_|-' in rstateid:
+                        st = salt.state.split_low_tag(rstateid)
+                    # in case of simple state, try to guess
+                    else:
+                        id_ = rstate.get('__id__', rstateid)
+                        if not id_:
+                            raise ValueError('no state id')
+                        st = {'__id__': id_, 'state': mod_}
+                    if mod_ == st['state'] and fstate == st['__id__']:
+                        ofresult = rstate.get('result', _empty)
+                        if ofresult in [False, True]:
+                            nret = ofresult
+                        if ofresult is False:
+                            # as soon as we find an errored onfail, we stop
+                            break
+                # consider that if we parsed onfailes without changing
+                # the ret, that we have failed
+                if nret is None:
+                    nret = False
+    return nret
+
+
+def check_result(running, recurse=False, highstate=None):
+    '''
+    Check the total return value of the run and determine if the running
+    dict has any issues
```

Comment:
This is a bit vague to me - any chance you can add a bit more description about what kinds of issues this checks for? (I'm curious how this differs from the `verify_ret_from_export` func in #43070, which I considered putting into `salt.utils.state`.)

Reply:
@aneeshusa That's a good question that I'll let @thatch45 answer, since he was the original author of this function.

This function is used to check against the state result in many places across salt, and this PR is just moving the function to a new home as part of an on-going effort between @terminalmage and I to break up the massive `salt/utils/__init__.py` file in to smaller, more useful pieces.

Reply:
Sure, I was just curious about this and probably best addressed in a follow-up, shouldn't block this PR going forward. Proceed with decomp!

Reply:
Yeah, sounds good. Thank you for your review, as always. It's greatly appreciated. I'll try to swing by your other PR that also add the state util file soon. :)